### ISSUE TITLE
Get another function's parameter types for use in another function's rest parameter

### ISSUE THREAD

**Issue by STofone** [microsoft/TypeScript#58991]
### 🔎 Search Terms

Rest parameter

### 🕗 Version & Regression Information

TypeScript Version 5.5.2

### 💻 Code

```ts
type AnyFunction = (...args: any[]) => any;
interface TestService {
  [key: string]: {
    [key: string]: AnyFunction;
  };
}
declare function f<S extends keyof TestService , F extends keyof TestService[S], A extends Parameters<TestService[S][F]>>(k1: S, k2: F, ...args: A): void;
```

### 🙁 Actual behavior

❌report at A : Rest parameter must be an array type or a generic with array constraint

### 🙂 Expected behavior

It can infer another function's params tuple as a rest param type

### Additional information about the issue

#29919 seems similar
found in stackoverflow https://stackoverflow.com/questions/76772628/get-another-functions-parameter-types-for-use-in-another-functions-rest-par

**Comment by MartinJohns** [comment]
You forgot to fill out the issue template completely.

The code does not show any error in the [playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAggdiAYgVzgY2ASwPZygXigAoA6MgQwCcBzAZwC4pyEBtAXQEoCA+JhAbgBQmOMAiUAZuTTQAKhFrAAyuIBumGVADegqFBYBrCCEaLKI6m0Y69ew8dPBzcS43hJUGHHCF6AvkJ+ggAmEGgANlTQEp5YuFASADxKUBAAHmJwwbRQRiDYElDyiiqU6poANFCIqRkQWTl5BUUKymoaECxKbFUwtZnZUAAKVOQAthBilLSJxW1lHV1sLIhs3NxEBgCMjEpVBgBMjIhVZCRUdG4cjKrYmMFCQA).

**Comment by STofone** [comment]
![image](https://github.com/microsoft/TypeScript/assets/37892335/5edef7d3-2233-44a9-bfc5-f9be4274e640)
same code in idea

**Comment by STofone** [comment]
![image](https://github.com/microsoft/TypeScript/assets/37892335/75e9a572-2c43-451a-92ea-05a1056ef6cf)
Weird difference between each way of function definition inside object

**Comment by MartinJohns** [comment]
> Weird difference between each way of function definition inside object

Properties and methods have different variance when `strictFunctionTypes` is enabled. And functions and lambdas have different `this` context.

**Comment by STofone** [comment]
> Properties and methods

[playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAggdiAYgVzgY2ASwPZygXigAoA6MgQwCcBzAZwC4pyEBtAXQEoCA+JhAbgBQgiAA8w2SsCiY4wCJQBm5NNAAqEWsADKCgG6ZVUAN6CoUFgGsIIRlsqzqbRqfPmrNu8AdwnjeEioGDhwQuYAvkLhwmISUjJyCspGGlq6lAZGrlAARJYAjDkuZm55AExFxORePtRc+Lx62JgAJmFQkYLRgmi4WlDyafqGEIypOsNGhNl5hcVuuZYV9CULRNVQ9o71vNkL5r1wtNgANhAkJ9jU6xyrbtERXcKH-YPS0yWK+YwAPNpQYnkcBatCg1hA2EUUHG6UyEAANFBEADRECQWCbJDoZoJhkRixtGxETAUWjQQAFKjkAC2EHklFoPxhkwgBLYLEQbG43CIBUY2kRS0YiERZBIVDo-h2JjulDpyEoeDesPxBXZSzYpAoNFotwi8M+ZT+pIgwNB4KxzLxqkRyMBpvRFqhVrhbOJJrNUEplBpdIUjJd+MJHK5PL5UAFYLKwtF2slsC4eygcuACqVOJVqis+XVZU1YolupK4QN5kUAGZGIoglhcFBjfbPU7sUNrQikR7HZjnRmWW7YJ2KVTafSA7222yQ9zed8I4Lo0jY+KdVKZQsU2mBuPXWqrHmtcu6HqOgbum8SF8iLMcojyjfcvkyuWcrdz4oyleCve77fH8-XziF7lp+hS3ks35-i+whAA)
I understand the different `this` context, but it seems has no difference of input param types.
